core/src/main/scala/kafka/tools/DumpLogSegments.scala

```diff
@@ -124,15 +134,73 @@ object DumpLogSegments {
         println("offset: %d position: %d".format(entry.offset + index.baseOffset, entry.position))
     }
   }
-  
+
+  private trait MessageParser {
+    def parse(message: Message): (Option[_], Option[_])
```

Comment:
This is a bit weird. One would expect `MessageParser` to take two type parameters, one for the key and the other for the payload. Probably OK for this PR since the trait is private and because of how we use `Decoder` (we assume that the instance returned by `Decoder` will have a reasonable `toString` so we ignore the type parameter of `Decoder`. This is not great, it would be better to use the type system make this requirement explicit).